File: frontend/src/types/sppb.ts
export interface SppbItem {
  id: string
  line_no: number
  description: string
  unit: string
  quantity_ordered: string
  quantity_delivered: string
}

export interface Sppb {
  id: string
  /** Contoh: 001/SPPB/VII/2026 — selalu dibuat server */
  number: string
  sequence_no: number
  year: number
  project_id: string
  project_code: string
  project_name: string
  vendor_id: string
  vendor_name: string
  issued_date: string
  notes: string | null
  items: SppbItem[]
  created_at: string
  updated_at: string
}

export interface SppbListQuery {
  page?: number
  size?: number
  project_id?: string
  vendor_id?: string
}

File: backend/src/app/service_layer/schemas/sppb.py
import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class SppbItemCreate(BaseModel):
    description: str = Field(..., min_length=1)
    unit: str = Field(..., max_length=30)
    quantity_ordered: Decimal = Field(..., gt=0)


class SppbItemRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    line_no: int
    description: str
    unit: str
    quantity_ordered: Decimal
    quantity_delivered: Decimal


class SppbCreate(BaseModel):
    project_id: uuid.UUID
    issued_date: date
    notes: str | None = None
    items: list[SppbItemCreate] = Field(..., min_length=1)


class SppbItemProgress(BaseModel):
    id: uuid.UUID
    quantity_delivered: Decimal = Field(..., ge=0)


class SppbProgressUpdate(BaseModel):
    items: list[SppbItemProgress] = Field(..., min_length=1)


class SppbRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    number: str
    sequence_no: int
    year: int
    project_id: uuid.UUID
    project_code: str
    project_name: str
    vendor_id: uuid.UUID
    vendor_name: str
    issued_date: date
    notes: str | None
    items: list[SppbItemRead] = []
    created_at: datetime
    updated_at: datetime
